### PR TITLE
added readme for multip-gpu training including ddp

### PR DESCRIPTION
**Pull request by wendywwang**
please review the content

### CODE REVIEW

File: Multi-GPU Training.md [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+**TL;DR**
+
+Use DP if you
+- are used to the way to train DeepLIIF on multiple GPUs since its first release, OR
+- do **not** have multiple GPU machines to utilize, OR
+- are fine with the training being a bit slower
+
+Use DDP if you
+- are willing to try a slightly different way to launch the training than before, OR
+- do have multiple GPU machines for cross-node distributino, OR
```
Comment (rich-nieto-ibm) [inline Multi-GPU Training.md]: typo: cross-node *distribution*

File: Multi-GPU Training.md [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+## Distributed Data Parallel (DDP)
+DDP usually spawns multiple processes. 
+
+**DeepLIIF's code follows the PyTorch recommendation to spawn 1 process per GPU** ([doc](https://github.com/pytorch/examples/blob/master/distributed/ddp/README.md#application-process-topologies)). If you want to assign multiple GPUs to each process, you will need to make modifications to DeepLIIF's code (see [doc](https://pytorch.org/tutorials/intermediate/ddp_tutorial.html#combine-ddp-with-model-parallelism)).
+
+
+### Train with DDP
+#### 1. Local Machine
+To launch training using DDP on a local machine, simply add `torchrun --nproc_per_node <gpu_num>` before the training command. Example with 2 GPUs (on 1 machine):
```
Comment (rich-nieto-ibm) [inline Multi-GPU Training.md:44]: We should probably mention that `torchrun` is available is torch >= 1.10.0

File: Multi-GPU Training.md [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    init_process()
+    subprocess.run(f'python cli.py train --dataroot {root_folder} --remote True --batch-size 3 --gpu-ids 0',shell=True)
+```
+Note that
+1. Always provide `--gpu-ids 0` to the training command for each process/pod if the gpu id gets re-named in each pod. If not, you will need to pass the correct gpu id in a dynamic way, possibly through an environment variable in each pod.
+
+#### 3. Multiple Virtual Machines
+To launch training across multiple VMs, you can refer to the scheduler framework you use. For each process, similar to the example for kubernetes, you will need to initiate the process group so that the current process knows who it is, where are its peers, etc., and then execute the regular training command in a subprocess.
+
+## Move from Single-GPU to Multi-GPU: Impact on Hypter-Parameters
```
Comment (rich-nieto-ibm) [inline Multi-GPU Training.md]: Typo Impact on *Hyperparameters*

File: Multi-GPU Training.md [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+To achieve equivalently good training results, you may want to adjust some hypter-parameters you figured out for a single GPU training.
```
Comment (rich-nieto-ibm) [inline Multi-GPU Training.md]: Typo: adjust some *hyperparameters*

File: Multi-GPU Training.md [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+### Batch Size & Learning Rate
+Backward propagation by default runs at the end of every batch to find how much changes to make in parameters. An immediate outcome from using multiple GPUs is that we have a larger effective batch size.
```
Comment (rich-nieto-ibm) [inline Multi-GPU Training.md]: Typo: how much *change*

File: Multi-GPU Training.md [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+## Move from Single-GPU to Multi-GPU: Impact on Hypter-Parameters
+To achieve equivalently good training results, you may want to adjust some hypter-parameters you figured out for a single GPU training.
+
+### Batch Size & Learning Rate
+Backward propagation by default runs at the end of every batch to find how much changes to make in parameters. An immediate outcome from using multiple GPUs is that we have a larger effective batch size. 
+
+In DDP, this means fewer gradient descent because DDP averages the gradients from all processes ([doc](https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html#torch.nn.parallel.DistributedDataParallel)). Assume thats in 1 epoch, a single-GPU training does gradient descent for 200 times. Now with 2 GPUs/processes, the training will have 100 batches in each process and does the gradient descent using the averaged gradients of the 2 GPUs/processes, one for each batch, which is 100 times.
```
Comment (rich-nieto-ibm) [inline Multi-GPU Training.md]: Typo: Assume *that* in 1 epoch

**Review by rich-nieto-ibm — CHANGES_REQUESTED** [review]
Just a few small typos, otherwise nice and succinct and clear.

File: Multi-GPU Training.md (hunk 1 above)
Comment (wendywwang) [reply Multi-GPU Training.md]: fixed

File: Multi-GPU Training.md (hunk 2 above)
Comment (wendywwang) [reply Multi-GPU Training.md:44]: added a note for this and the alternative command if using older versions

File: Multi-GPU Training.md (hunk 3 above)
Comment (wendywwang) [reply Multi-GPU Training.md]: fixed

File: Multi-GPU Training.md (hunk 4 above)
Comment (wendywwang) [reply Multi-GPU Training.md]: fixed

File: Multi-GPU Training.md (hunk 5 above)
Comment (wendywwang) [reply Multi-GPU Training.md]: fixed

File: Multi-GPU Training.md (hunk 6 above)
Comment (wendywwang) [reply Multi-GPU Training.md]: fixed

**Comment by wendywwang** [comment]
@rich-nieto-ibm updated the readme based on the comments and unresolved the conversations - feel free to unresolve them if you see any change is still not good enough!

**Review by rich-nieto-ibm — APPROVED** [review]

**Comment by wendywwang** [comment]
@rich-nieto-ibm @drewmibm please review the additional section added for visualizer

**Review by rich-nieto-ibm — APPROVED** [review]
Oops didn't mean to retract approval!

**Review by rich-nieto-ibm — APPROVED** [review]
I'm curious what MSK will say about Visdom portion as it really only impacts them (because they'll be using CPD). The Visdom portion makes sense to me, but I imagine if I was first time reader I'd be a little lost. 

Another option that I saw that could make this a little simpler involves averaging the losses from the different workers and then only visualizing the summarized loss data. I dunno about the images, though... I'll dig through and fine the link I found showing how to do this. It's just a few lines of code which performs an all_reduce step to sum the losses and then you di
vide by # workers you then add a dist.barrier() call to resync the workers.

**Comment by wendywwang** [comment]
resolved the above comments offline
